Edits and Replies support

lotusronin opened the pull request:
Fixes #43, #44, #61
Adds support for sending replies to messages and editing messages the user has sent in the swing gui.

Edits only work on text type messages right now, and some work should probably be done on better fallback messages for the replies, and both could be rendered better on the gui.

I would also eventually like to make replies have a clickable link portion that scrolls to the message it is relating to.

lotusronin committed:
Initial reply support.

lotusronin committed:
Merge branch 'audio_message_support' into edits_replies_support

lotusronin committed:
Add support for editing messages.

Minor changes to replies.

lotusronin committed:
Cleanup rendering edits and combine buttons as a menu.

lotusronin committed:
Add view source option for message events

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Events.kt:
```diff
@@ -19,10 +19,17 @@ data class LoginIdentifier(val type: String, val user: String)
+data class SendRoomMessage(
+    val msgtype: String,
+    val body: String,
+    @SerialName("m.relates_to") val replybock: RelationBlock? = null
```
It seems odd to combine regular message sending with replying, but not with editing. I feel like they could either be all combined or all separate - my gut leans towards combined to reduce duplication, but I'm not sure.

lotusronin replied:
Combined the classes for now but a look at the sending code architecture is likely warranted.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Events.kt:
```diff
@@ -136,6 +168,30 @@ class RoomEventFallback(
+)
+
+@Serializable
+class SendMessageEdit(
```
Other half of the all combined / all separate question. It looks like these can all be combined ok, no bad name overlap or anything. We could always do an inheritance thing, but I'm pretty sure that would be overkill.

lotusronin replied:
Combined the two classes.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt:
```diff
@@ -66,13 +66,20 @@ class MatrixSession(val client: HttpClient, val access_token: String, var transa
     fun getRoomEvents(id: String) = synchronized(this) { sync_response!!.rooms.join[id]!!.timeline.events }
 
-    fun sendMessage(msg: String, room_id: String): Outcome<String> {
+    fun sendMessage(msg: String, room_id: String, reply_id: String = ""): Outcome<String> {
```
Ditto combine vs separate thing happening here. I suppose this is part of a bigger discussion with sendImageMessage, etc. Maybe they should all have separate entrance points and all call a shared method?

lotusronin replied:
I don't mind all of the entrance points calling a shared method. Perhaps make the shared function generic on request body type, though I'm not a huge fan of that right now.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt:
```diff
@@ -78,44 +78,120 @@ class MatrixRooms(private val msession: MatrixSession, val message: String) : Ma
+    open val id: String,
+    open val timestamp: Long,
+    open val replied_event: String = "",
+    open val edited_event: String = ""
```
Perhaps these should be pointers to the SharedUiMessage they refer to?

lotusronin replied:
Not sure it makes sense to have edited_event be a pointer as it was just used to specifically mark the message with an "(edited)" tag and in the edits map construction. Replies it makes more sense, though what if we have a timeline state with the reply but not the original message in memory? I was going to use the field to find the message we want to jump to on message clicks in the future once we got some progress on view ports.

Limvot replied:
That's fair, that makes sense

lotusronin replied:
I did remove the now useless edited_event field though.

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt:
```diff
@@ -78,44 +78,120 @@ class MatrixRooms(private val msession: MatrixSession, val message: String) : Ma
+) : SharedUiMessage(sender,message,id,timestamp)
+
 class MatrixChatRoom(private val msession: MatrixSession, val room_id: String, val name: String) : MatrixState() {
+    val edits: MutableMap<String,SharedUiMessage> = mutableMapOf()
```
I think the shared UI should do the edit combination for the UI, not exposing the mutable edit map, just the same list of messages it always has, just pre-edited. This would keep the UI thinner, and it can use extra fields on the SharedUiMessage to display it specially or allow jumping to the referenced event. Gives edit handling for free for simple GUIs too!

lotusronin replied:
Agreed

Limvot commented on serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/States.kt:
```diff
@@ -78,44 +78,120 @@ class MatrixRooms(private val msession: MatrixSession, val message: String) : Ma
+        for(event in msession.getRoomEvents(room_id)) {
+            if (event as? RoomMessageEvent != null) {
+                if(event.event_id == msg_id) {
+                    return event.raw_self.toString()
```
Hehehe, there has to be a way to get the pretty printed JSON, though I admit I didn't find it with some quick searching...

lotusronin replied:
I was so frustrated trying to find an easy way. The only way I found readily available was string -> obj -> specialized format object from serialization lib -> back to string and that seemed way too overkill for the debug feature.

Limvot commented on serif_swing/src/main/kotlin/serif_swing/App.kt:
```diff
@@ -160,7 +191,21 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
             message_field.text = ""
-            transition(m.sendMessage(text), true)
+            val res =
+            if(replied_event_id == "" && edited_event_id == "") {
```
I think you could use a when expression to pattern match on the cases, though I'm not positive.

lotusronin replied:
You can indeed.

Limvot commented on serif_swing/src/main/kotlin/serif_swing/App.kt:
```diff
@@ -12,6 +12,32 @@ import javax.swing.*
+    }
+}
+
+var logged_in_user: String = ""
```
I think the currently logged in user should be retrieved from MatrixSession, as I think it's returned from login

lotusronin replied:
Also agree, just didn't feel like scouring the MatrixSession data for it the night I implemented this stuff.

Limvot commented on serif_swing/src/main/kotlin/serif_swing/App.kt:
```diff
@@ -208,20 +255,97 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
+                val w = dim.width
+                val dialog = JDialog(window, "Event Source")
+
+                val top_panel = JPanel()
```
I think this panel doesn't do anything, is it necessary?

lotusronin replied:
Good call, we can indeed get rid of it

Limvot reviewed (changes requested):
Very cool stuff! I do have some architecture suggestions, but solid overall.

lotusronin committed:
Filter messages with edits in shared code.

Also remove unnecessary panel.

lotusronin committed:
Address some more PR feedback.

Pull username from shared code. Restructure an if else chain as a when
block.

lotusronin committed:
Remove unused field from SharedUiMessage.

lotusronin committed:
Merge send Room Message and Edit event classes.

lotusronin committed:
Clean up message sending a bit.

lotusronin committed:
Fixes for PR feedback.

Remove image terms from media cache. Only show edits for messages if the
senders match.

lotusronin committed:
Merge branch 'main' into edits_replies_support
